Add remaining methods to SearchServiceClient

heaths opened the pull request:
An initial look at the API changes. I'll add a recorded hero test or two in another commit to add to this PR, but wanted to get the brunt of the impactful work ready for review.

heaths committed:
Add remaining methods to SearchServiceClient

tg-msft commented on sdk/search/Azure.Search.Documents/src/SearchServiceClient.cs:
```diff
@@ -261,30 +256,309 @@ public virtual SearchIndexClient GetSearchIndexClient(string indexName)
+        [ForwardsClientCalls]
+        public virtual Response<DataSource> CreateOrUpdateDataSource(
+            DataSource dataSource,
+            MatchConditions accessConditions = null,
+            SearchRequestOptions options = null,
```
We might consider creating an Options class that unions these together for this operation.  We can address that comprehensively in the future.

heaths replied:
It had crossed my mind. I'll open a bug to consider.

heaths replied:
Opened #11052 to track.

tg-msft commented on sdk/search/Azure.Search.Documents/src/SearchServiceClient.cs:
```diff
@@ -261,30 +256,309 @@ public virtual SearchIndexClient GetSearchIndexClient(string indexName)
+        /// <exception cref="RequestFailedException">Thrown when a failure is returned by the Search service.</exception>
+        [ForwardsClientCalls]
+        public virtual Response<IReadOnlyList<DataSource>> GetDataSources(
+            IEnumerable<string> select = null,
```
I wonder if we should give these a name like `selectedFields` or something?  Doesn't have to happen as part of this PR.

heaths replied:
Not a bad idea. The rest would be, based on docs, `selectedProperties`. Seems odd that one would be different. I think I might just standardize on "properties" for consistency, though.

heaths replied:
I ended up going with `selectProperties` - more a verb form that matches the intent. I could find no solid guideline or preexisting art in our own APIs, but .NET guidelines did mention naming with intent (as oppose to effect). Not married to it, though. Let me know if you would rather have it be `selectedProperties`. I also just considered `properties`, which is fairly consistent in various parts of the .NET BCL, but I figured that's only slightly less ambiguous than `select`; but, I'm also not opposed to that.

tg-msft replied:
That sounds good to me.  I think we'll spend some time early next preview looking at the entire surface in ApiView and maybe one of these options will feel like a clear winner.

brjohnstmsft replied:
@heaths @tg-msft The main motivation for allowing `$select` for these APIs was to support the scenario where you just list the names of resources without the rest of the objects. So one option would be to expose `GetXYZNames` APIs instead of having this property.

tg-msft commented on sdk/search/Azure.Search.Documents/src/SearchServiceClient.cs:
```diff
@@ -330,82 +593,1058 @@ public virtual SearchIndexClient GetSearchIndexClient(string indexName)
+        /// <returns>The <see cref="Response{T}"/> from the server containing a list of <see cref="SynonymMap"/>.</returns>
+        /// <exception cref="RequestFailedException">Thrown when a failure is returned by the Search service.</exception>
+        [ForwardsClientCalls]
+        public virtual Response<IReadOnlyList<SynonymMap>> GetSynonymMaps(
```
Should we make all these `Get*s` `Pageable<T>`?  We can discuss it after this Preview with Bruce to get an idea of how these features might grow over time.

heaths replied:
I opened #11053 to track, but the models themselves aren't currently pageable, and I imagine they'd never yield enough results to bother making them that way. As you said, we can discuss later.

tg-msft commented on sdk/search/Azure.Search.Documents/src/SearchServiceClient.cs:
```diff
@@ -193,27 +202,20 @@ public virtual SearchIndexClient GetSearchIndexClient(string indexName)
         /// This operation returns the number and type of objects in your
         /// service, the maximum allowed for each object type given the service
         /// tier, actual and maximum storage, and other limits that vary by
-        /// tier.  This request pulls information from the service so that you
+        /// tier. This request pulls information from the service so that you
```
I consider doc comments to be a whitespace sensitive language and this is a syntax error.

_(+@jsquire and @annelo-msft  to back me up here...)_

heaths replied:
I know emacs isn't a far cry from a typewriter 😉, but the XML doc to HTML and/or markdown renderer canonicalizes it *properly* anyway.

tg-msft replied:
_Ohhh..._  I didn't realize that Vim uses a variable width font.  That's cute.  Two spaces would render the same in our ref docs, but I thought you were using a real editor where the additional space was a useful affordance.  😄

tg-msft reviewed (approved):
Looks great!

heaths committed:
Resolve PR feedback and update API

heaths committed:
Add tests for preview 2

More tests will be added for preview 3, but these tests exercise a decent amount of the new SearchServiceClient APIs in a plausible use case.

tg-msft commented on sdk/search/Azure.Search.Documents/tests/Utilities/SearchResources.cs:
```diff
@@ -442,6 +547,90 @@ private async Task<SearchResources> CreateSearchServiceIndexAndDocumentsAsync()
+
+                foreach (Hotel hotel in hotels)
+                {
+                    Task task = Task.Run(async () =>
+                    {
```
The body looks like it should be fine directly in the `foreach`.

heaths replied:
I don't want to execute them now in sequence. That's why create and add a task to the list to execute in parallel below in `await Task.WhenAll`. Faster this way, or is there some batching you're doing in the SDK that will do this in bulk already?

tg-msft replied:
I see - makes sense.

tg-msft commented on sdk/search/Azure.Search.Documents/tests/Utilities/SearchResources.cs:
```diff
@@ -442,6 +547,90 @@ private async Task<SearchResources> CreateSearchServiceIndexAndDocumentsAsync()
+        /// Creates a new storage account.
+        /// </summary>
+        /// <returns>The current <see cref="SearchResources"/>.</returns>
+        private async Task<SearchResources> CreateStorageAccountAsync()
```
Does this have to happen per test?  Or could we do this the "normal" way with our test-resources.json and use a random container inside the storage account?

heaths replied:
It was done following the same pattern we do for Search services. I suppose I could create a single storage account using test-resources.json, but 1) that doesn't get run for user-driven tests, and 2) resolving that creates an inconsistency with Search service. We'd still have to create one in the F5/test run cases and only one, but could otherwise randomize containers - unless we can safely assume the container will be immutable. Thinking about that more, that problably is a safe assumption.

Was there some concern? Creating these resources is pretty fast.

tg-msft replied:
I would just prefer to avoid management plane dependencies in our tests.  We couldn't for the Search service instances.  If it's possible to do that for Storage, I would prefer to and maintain as much adherence to the unified test-resources.json approach as possible.  We'd have to create one ahead of time for local testing, but we already have to do that for the resource group/principal/etc.  We'd just have another env var with the connection string.

heaths replied:
We'll need to update the contrib docs to tell people to use it beforehand. I've opened #1108 to track. We should use that to create the resource group (e.g. "rg-$BaseName") and storage account.

tg-msft reviewed (approved):
Looks good - left a few comments on the test infrastructure changes, but they don't need to be addressed as part of this PR.

tg-msft commented:
/azp run net - search - ci

heaths committed:
Update Search APIs
